sys/net/network_layer/ng_sixlowpan/frag/rbuf.c

```diff
@@ -0,0 +1,307 @@
+static rbuf_int_t *_rbuf_int_get_free(void);
+/* remove entry from reassembly buffer */
+static void _rbuf_rem(rbuf_t *entry);
+/* update interval buffer of entry */
+static bool _rbuf_update_ints(rbuf_t *entry, uint16_t offset, size_t frag_size);
+/* checks timeouts and removes entries if necessary (oldest if full) */
+static void _rbuf_gc(void);
+/* gets an entry identified by its tupel */
+static rbuf_t *_rbuf_get(const void *src, size_t src_len,
+                         const void *dst, size_t dst_len,
+                         size_t size, uint16_t tag);
+
+void rbuf_add(ng_pktsnip_t *pkt)
+{
+    ng_netif_hdr_t *hdr = pkt->next->data;
+    rbuf_t *entry;
+    ng_sixlowpan_frag_t *frag = pkt->next->data;
+    size_t datagram_size;
+    size_t frag_size;
+    uint16_t tag, offset = 0;
+    kernel_pid_t iface = hdr->if_pid;
+    uint8_t *data;
+    bool first;
+
+    switch (frag->disp_size.u8[0] & NG_SIXLOWPAN_FRAG_DISP_MASK) {
+        case NG_SIXLOWPAN_FRAG_1_DISP:
+            tag = byteorder_ntohs(frag->tag);
+            datagram_size = byteorder_ntohs(frag->disp_size) & NG_SIXLOWPAN_FRAG_SIZE_MASK;
+            frag_size = (pkt->size - sizeof(ng_sixlowpan_frag_t));
+            data = (uint8_t *)(hdr + 1);
+            first = true;
+            break;
+
+        case NG_SIXLOWPAN_FRAG_N_DISP:
+            tag = byteorder_ntohs(frag->tag);
+            datagram_size = byteorder_ntohs(frag->disp_size) & NG_SIXLOWPAN_FRAG_SIZE_MASK;
+            offset = (((ng_sixlowpan_frag_n_t *)frag)->offset * 8);
+            frag_size = (pkt->size - sizeof(ng_sixlowpan_frag_n_t));
+            data = (uint8_t *)(hdr + 1) + 1;
+            break;
+
+        default:
+            DEBUG("6lo rbuf: Not a fragment header.\n");
+            ng_pktbuf_release(pkt);
+
+            return;
+    }
+
+    _rbuf_gc();
+    entry = _rbuf_get(ng_netif_hdr_get_src_addr(hdr), hdr->src_l2addr_len,
+                      ng_netif_hdr_get_dst_addr(hdr), hdr->dst_l2addr_len,
+                      datagram_size, tag);
+
+    if (entry == NULL) {
+        DEBUG("6lo rbuf: reassembly buffer full.\n");
+        ng_pktbuf_release(pkt);
+
+        return;
+    }
+
+    if (_rbuf_update_ints(entry, offset, frag_size)) {
+        if (first) {
+            /* some dispatches do not count to datagram size and we need
+             * more space because of that */
+            switch (data[0]) {
+                case NG_SIXLOWPAN_UNCOMPRESSED:
+                    if (ng_pktbuf_realloc_data(pkt, pkt->size + 1) < 0) {
+                        DEBUG("6lo rbuf: could not reallocate packet data.\n");
+                        ng_pktbuf_release(pkt);
+                        return;
+                    }
+
+                    /* move already inserted fragments 1 to the right */
+                    for (int i = pkt->size - 1; i > 0; i--) {
+                        uint8_t *d = ((uint8_t *)(pkt->data)) + i;
+                        *d = *(d - 1);
```

Comment:
> That was not my statement. For more complex memory movement or even just
> for more than one byte we need to reevaluate this.

Yes, and we might even want to change the libc implementation at some point,
but since we don't have a generic `memmove()` replacement and here and in
@jremmert-phytec-iot's PR the proposed alternative seems more efficient, the
solution is fine.